server/server.go

```diff
@@ -69,36 +74,74 @@ func (s *Server) prep() {
 //
 // Specifying one of CertFile/KeyFile without the other will panic.
 func (s *Server) ListenAndServe() error {
+	if s.HandleOSSignals {
+		return s.listenAndServeHandleOSSignals()
+	}
+
+	return s.listenAndServe()
+}
+
+// ListenAndServeTLS sets KeyFile and CertFile, then calls ListenAndServe
+func (s *Server) ListenAndServeTLS(certFile, keyFile string) error {
+	if len(certFile) == 0 || len(keyFile) == 0 {
+		panic("either CertFile/KeyFile must be blank, or both provided")
+	}
+	s.KeyFile = keyFile
+	s.CertFile = certFile
+	return s.ListenAndServe()
+}
+
+// Shutdown will gracefully shutdown the server, using a default shutdown
+// timeout if a context is not provided.
+func (s *Server) Shutdown(ctx context.Context) error {
+
+	if ctx == nil {
+		ctx, _ = context.WithTimeout(context.Background(), s.DefaultShutdownTimeout)
+	}
+
+	return s.Server.Shutdown(ctx)
+}
+
+func (s *Server) listenAndServe() error {
+
 	s.prep()
+	if len(s.CertFile) > 0 || len(s.KeyFile) > 0 {
+		return s.Server.ListenAndServeTLS(s.CertFile, s.KeyFile)
+	}
+
+	return s.Server.ListenAndServe()
+}
+
+func (s *Server) listenAndServeHandleOSSignals() error {
 	stop := make(chan os.Signal, 1)
 	signal.Notify(stop, os.Interrupt, os.Kill)
 
+	s.prep()
 	if len(s.CertFile) > 0 || len(s.KeyFile) > 0 {
-		if len(s.CertFile) == 0 || len(s.KeyFile) == 0 {
-			panic("either CertFile/KeyFile must be blank, or both provided")
-		}
-		go func() {
-			if err := s.Server.ListenAndServeTLS(s.CertFile, s.KeyFile); err != nil {
-				log.Error(err, nil)
-				os.Exit(1)
-			}
-		}()
+		s.listenAndServeTLSAsync(s.CertFile, s.KeyFile)
 	} else {
-		go func() {
-			if err := s.Server.ListenAndServe(); err != nil {
-				log.Error(err, nil)
-				os.Exit(1)
-			}
-		}()
+		s.listenAndServeAsync()
 	}
 
 	<-stop
-	return s.Shutdown(nil)
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+	return s.Shutdown(ctx)
 }
 
-// ListenAndServeTLS sets KeyFile and CertFile, then calls ListenAndServe
-func (s *Server) ListenAndServeTLS(certFile, keyFile string) error {
-	s.KeyFile = keyFile
-	s.CertFile = certFile
-	return s.ListenAndServe()
+func (s *Server) listenAndServeAsync() {
+	go func() {
+		if err := s.Server.ListenAndServe(); err != nil {
+			log.Error(err, nil)
+			os.Exit(1)
+		}
+	}()
+}
+
+func (s *Server) listenAndServeTLSAsync(certFile, keyFile string) {
```

Comment:
`certFile` and `keyFile` parameters aren't used (using the ones on `s` is probably right)

Reply:
You're right, these parameters were left over by mistake when refactoring. I have removed them.